src/goto-instrument/contracts/dynamic-frames/dfcc_loop_utils.cpp

```diff
@@ -0,0 +1,514 @@
+  {
+    auto &loop_content = loop_head_and_content.second;
+    if(loop_content.empty())
+      continue;
+
+    auto loop_head = loop_head_and_content.first;
+    auto loop_end = loop_head;
+
+    // Find the last back edge to `loop_head`
+    for(const auto &t : loop_content)
+    {
+      if(
+        t->is_goto() && t->get_target() == loop_head &&
+        t->location_number > loop_end->location_number)
+        loop_end = t;
+    }
+
+    if(loop_end == loop_head)
+    {
+      log.error() << "Could not find end of the loop starting at: "
+                  << loop_head->source_location() << messaget::eom;
+      throw 0;
+    }
+
+    // After loop-contract instrumentation, jumps to the `loop_head` will skip
+    // some instrumented instructions. So we want to make sure that there is
+    // only one jump targeting `loop_head` from `loop_end` before loop-contract
+    // instrumentation.
+    // Add a skip before `loop_head` and let all jumps (except for the
+    // `loop_end`) that target to the `loop_head` target to the skip
+    // instead.
+    insert_before_and_update_jumps(
+      function_body, loop_head, goto_programt::make_skip());
+    loop_end->set_target(loop_head);
+
+    // There should be only one latch node.
+    for(auto &t : loop_content)
+    {
+      if(
+        t->is_goto() && t->get_target() == loop_head &&
+        t->location_number < loop_end->location_number)
+      {
+        t->set_target(loop_end);
+      }
+    }
+
+    // Convert conditional latch node to unconditional latch node.
+    //       IF g THEN GOTO HEAD
+    // --------------------------
+    //       IF !g THEN GOTO EXIT
+    //       GOTO HEAD
+    // EXIT: skip
+    if(loop_end->has_condition() && !loop_end->condition().is_true())
+    {
+      const auto &exit =
+        function_body.insert_after(loop_end, goto_programt::make_skip());
+
+      insert_before_and_update_jumps(
+        function_body,
+        loop_end,
+        goto_programt::make_goto(
+          exit, not_exprt(loop_end->condition()), loop_end->source_location()));
+
+      exprt loop_end_condition = loop_end->condition();
+      loop_end_condition.set(ID_value, ID_true);
+      *loop_end = goto_programt::make_goto(
+        loop_head, loop_end_condition, loop_end->source_location());
+    }
+
+    // By definition the `loop_content` is a set of instructions computed
+    // by `natural_loops` based on the CFG.
+    // Since we perform assigns clause instrumentation by sequentially
+    // traversing instructions from `loop_head` to `loop_end`,
+    // here we ensure that all instructions in `loop_content` belong within
+    // the [loop_head, loop_end] target range
+
+    // Check 1. (i \in loop_content) ==> loop_head <= i <= loop_end
+
+    // Check 2. no jumps into instructions in loops from outside.
+    for(const auto &i : loop_content)
+    {
+      if(std::distance(loop_head, i) < 0 || std::distance(i, loop_end) < 0)
```

Comment:
How does this actually work? There is a lot of use of `auto` here (too much, IMHO) so I am guessing types: aren't `loop_head` and `i` std::list iterators? So the only way to compute the distance is increment-and-test. I don't see how a value of less-than-zero would be possible in that case?